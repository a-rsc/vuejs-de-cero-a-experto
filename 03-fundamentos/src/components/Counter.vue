<template>
    <div>
        <h3>Counter - {{ title || 'valor por defecto'}}</h3>
        <h3 data-testid="title">Counter - {{ CustomTitle }}</h3>
        <p style="text-align: justify;">Mientras se edita el template se mantiene el estado,es decir, el valor de la variable no se modifica.</p>
        <p data-testid="counter">{{ counter }}</p>
        <p>{{ counter }}<sup>2</sup> = {{ getSquareValue() }} (no guardado en memoria)</p>
        <p>{{ counter }}<sup>2</sup> = {{ getSquareValue() }} (no guardado en memoria)</p>
        <p>{{ counter }}<sup>2</sup> = {{ getSquareValue() }} (no guardado en memoria)</p>
        <p>{{ counter }}<sup>2</sup> = {{ getSquareValue() }} (no guardado en memoria)</p>

        <p>{{ counter }}<sup>2</sup> = {{ SquareCounter }} (guardado en memoria - propiedad computada)</p>
        <p>{{ counter }}<sup>2</sup> = {{ SquareCounter }} (guardado en memoria - propiedad computada)</p>
        <p>{{ counter }}<sup>2</sup> = {{ SquareCounter }} (guardado en memoria - propiedad computada)</p>
        <p>{{ counter }}<sup>2</sup> = {{ SquareCounter }} (guardado en memoria - propiedad computada)</p>

        <div class="buttons">
            <button @click="AddCounter">+ 1</button>
            <button @click="SubtractCounter">- 1</button>
            <button @click="ResetCounter">Reset</button>
        </div>
    </div>
</template>

<script>
export default {
    props: {
        title: String,
        start: {
            type: Number,
            required: false,
            default: 30,
            // https://v3.vuejs.org/guide/component-props.html#prop-validation
            validator(value) {
                return value > 10
            }
        }
    },
    name: 'Snipper',
    data() {
        return {
            counter: this.start
        }
    },
    methods: {
        AddCounter() {
            this.counter += 1
        },
        SubtractCounter() {
            this.counter -= 1
        },
        ResetCounter() {
            this.counter = this.start
        },
        getSquareValue() {
            // console.log('getSquareValue()')
            return this.counter * this.counter
        }
    },
    computed: {
        CustomTitle() {
            // return this.title ?? 'valor por defecto'
            return this.title || 'valor por defecto'
        },
        // Propiedad que se maneja en cache (memoria)
        SquareCounter() {
            // console.log('SquareCounter()')
            return this.counter * this.counter
        }
    }
}
</script>

<style>
button {
    /* ordenar l√≠neas ascendente */
    background-color: #64B687;
    border-radius: 5px;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
}
button:hover {
    background-color: #5aa67b;
    transition: 0.3s ease-in-out;
}
</style>