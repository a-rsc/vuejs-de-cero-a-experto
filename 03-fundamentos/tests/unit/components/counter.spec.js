// npm run test:unit counter
// Ha aparecido una carpeta __snapshots___
// Si cambiamos se producirá un error

// Si el cambio es nuevo -> npm run test:unit -u counter

// mount -> monta todo el componente
// shallowMount -> parecido pero menos algunas propiedades y funciones...
// import { mount } from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter'

// Material adjunto
// https://jestjs.io/docs/setup-teardown
describe('Counter Component', () => {
/*
  // la función it es muy parecida
  test('Debe de hacer match con el snapshot', () => {

    // AAAA
    // Arreglar
    const wrapper = shallowMount(Counter)

    // Estímulo

    // Aserción
    expect( wrapper.html() ).toMatchSnapshot()
  })
*/

  // let wrapper;

  // beforeEach(() => {
  //   wrapper = shallowMount(Counter) // aplicación que corre independientemente
  // })

  test('h3 debe contener "Counter - valor por defecto"', () => {

    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    expect(wrapper.find('h3').exists()).toBeTruthy()

    const h3 = wrapper.find('h3')

    // console.log(h3.text())

    expect(h3.text()).toBe('Counter - valor por defecto')
  })

  test('el valor por defecto debe de ser 100 en el p', () => {
/*
    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    const pTags = wrapper.findAll('p')

    // console.log(pTags[1].text())

    expect(pTags[1].text()).toBe('100')

*/
  })

  test('el elemento title debe contener "Counter - valor por defecto"', () => {

    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    const h3 = wrapper.find('[data-testid="title"]')

    expect(h3.text()).toBe('Counter - valor por defecto')
  })

  // Hay que esperar que el DOM se actualice, necesitamos el async y await
  test('debe de incrementar y decrementar el contador', async() => {

    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    const buttons = wrapper.findAll('button')

    // Hay que esperar que el DOM se actualice
    await buttons[0].trigger('click') // +1

    let value = wrapper.find('[data-testid="counter"]')

    expect(value.text()).toBe('31')

    // Hay que esperar que el DOM se actualice
    await buttons[1].trigger('click') // -1
    await buttons[1].trigger('click') // -1

    expect(value.text()).toBe('29')
  })

  // Hay que esperar que el DOM se actualice, necesitamos el async y await
  test('debe de incrementar y decrementar el contador', async() => {

    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    const [increaseBtn, decreaseBtn] = wrapper.findAll('button')

    // Hay que esperar que el DOM se actualice
    await increaseBtn.trigger('click') // +1
    await increaseBtn.trigger('click') // +1
    await decreaseBtn.trigger('click') // -1
    await decreaseBtn.trigger('click') // -1
    await decreaseBtn.trigger('click') // -1

    const value = wrapper.find('[data-testid="counter"]').text()

    expect(value).toBe('29')
  })

  test('debe de establecer el valor por defecto', () => {

    const wrapper = shallowMount(Counter) // aplicación que corre independientemente

    // console.log(wrapper.props())
    // const {start} = wrapper.props()
    // console.log(start)

    const start = wrapper.props('start')
    // console.log(typeof start)

    const value = wrapper.find('[data-testid="counter"]').text()

    // Comprobación de tipos
    expect(Number(value)).toBe(start)
  })

  test('debe de mostrar la prop title', () => {

    const title = 'Hola mundo!!!!!'

    const wrapper = shallowMount(Counter, {
      props: {
        title: title
      }
    })

    expect(wrapper.find('h3').text()).toBe(`Counter - ${title}`)
  })
})