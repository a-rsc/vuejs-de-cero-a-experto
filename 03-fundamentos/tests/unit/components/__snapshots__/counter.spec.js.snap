// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Debe de hacer match con el snapshot 1`] = `
<div>
  <h3>Counter - valor por defecto</h3>
  <h3>Counter - valor por defecto</h3>
  <p style="text-align: justify;">Mientras se edita el template se mantiene el estado,es decir, el valor de la variable no se modifica.</p>
  <p>30</p>
  <p>30<sup>2</sup> = 900 (no guardado en memoria)</p>
  <p>30<sup>2</sup> = 900 (no guardado en memoria)</p>
  <p>30<sup>2</sup> = 900 (no guardado en memoria)</p>
  <p>30<sup>2</sup> = 900 (no guardado en memoria)</p>
  <p>30<sup>2</sup> = 900 (guardado en memoria - propiedad computada)</p>
  <p>30<sup>2</sup> = 900 (guardado en memoria - propiedad computada)</p>
  <p>30<sup>2</sup> = 900 (guardado en memoria - propiedad computada)</p>
  <p>30<sup>2</sup> = 900 (guardado en memoria - propiedad computada)</p>
  <div class="buttons"><button>+ 1</button><button>- 1</button><button>Reset</button></div>
</div>
`;
